Initialized MultiFactorDB & LinkMapDB classes to interface with pre-defined databases.

thomazbastiaanssen opened the pull request:
Hi! 
Here's some code to describe what my intent is :) 

First, let's take a look at ariadne:::ChocoPhlAn, as an example structure we can store in ariadne

```

ChocoPhlAn = list(
    eggnog2uniref50    = data.frame(eggnog   = factor(), uniref50 = factor()),
    eggnog2uniref90    = data.frame(eggnog   = factor(), uniref90 = factor()),
    go2uniref50        = data.frame(go       = factor(), uniref50 = factor()),
    go2uniref90        = data.frame(go       = factor(), uniref90 = factor()),
    ko2uniref50        = data.frame(ko       = factor(), uniref50 = factor()),
    ko2uniref90        = data.frame(ko       = factor(), uniref90 = factor()),
    level4ec2uniref50  = data.frame(ec       = factor(), uniref50 = factor()),
    level4ec2uniref90  = data.frame(ec       = factor(), uniref90 = factor()),
    pfam2uniref50      = data.frame(pfam     = factor(), uniref50 = factor()),
    pfam2uniref90      = data.frame(pfam     = factor(), uniref90 = factor()),
    uniref502uniref90  = data.frame(uniref50 = factor(), uniref90 = factor())
)
```

Now we can interact with the db as if it were a MultiFactor:
```
# generate a MultiFactorDB
x <- MultiFactorDB(ChocoPhlAn)
x

# Extract one underlying LinkMapDB object
x[["ko2uniref90"]]
x[["ko2uniref90"]] |> class()

# Turn the LinkMapDB into an actual LinkMap:
y <- as.LinkMap(x[["ko2uniref90"]])
y 
y |> class()
```

thomazbastiaanssen committed:
Initialized MultiFactorDB & LinkMapDB classes to interface with defined databases.

thomazbastiaanssen committed:
cleanup

RiboRings commented on R/LinkMapDB-methods.R:
```diff
@@ -0,0 +1,59 @@
+
+#' @export
+#'
+S7::`method<-`(MultiFactor::as.LinkMap, LinkMapDB, function(x) {
```
Can we have this as `S7::method(..., function ...` to follow the rest of the package?

RiboRings commented on R/LinkMapDB-methods.R:
```diff
@@ -0,0 +1,59 @@
+
+    x <- `names<-`(
+        data.frame(
+            id.x = rep(keys, vapply(values, length, 1L)),
```
`lengths` can be used instead of `vapply(values, length, 1L)`

RiboRings commented on R/AllClasses.R:
```diff
@@ -0,0 +1,97 @@
+    }
+)
+
+ChocoPhlAn = list(
```
It would be preferable to provide an example with the real ChocoPhlAn or Woltka databases, instead of including example structures in the package.

RiboRings commented on R/AllClasses.R:
```diff
@@ -0,0 +1,97 @@
+#' @seealso [MultiFactorDB()]
+#' @export
+#'
+LinkMapDB <- S7::new_class(
```
Just wondering, is LinkMapDB necessary or is MultiFactorDB enough?

thomazbastiaanssen replied:
Hm, agreed. LinkMapDB would have my vote, as each object corresponds neatly to one download

RiboRings commented on R/LinkMapDB-methods.R:
```diff
@@ -0,0 +1,59 @@
+    # Extract values
+    values <- lapply(line.content, FUN = function(x) x[-1])
+
+    x <- `names<-`(
```
For clarity I would separate `data.frame` and `names` commands as two different lines.

RiboRings commented on R/LinkMapDB-methods.R:
```diff
@@ -0,0 +1,59 @@
+#' # Extract one underlying LinkMapDB object
+#' LM.db <- MF.db[["ko2uniref90"]]
+#'
+#' LM <- MultiFactor::as.LinkMap(LM.db)
```
I am not sure exactly how to fit `importMapping` into this workflow, but I would prefer it over importing data with `as.LinkMap`, since the latter is not a self-explanatory name.

RiboRings reviewed (changes requested):
Thanks for the PR! It looks very nice.

Here a few points:

1. Instead of working directly with internal classes (LinkMapDB, MultiFactorDB), could we let users interact with those through functions? For example, `queryMapping("ChocoPhlAn", ...)` which gives a MultiFactorDB to easily navigate a database, and then `importMapping("ChocoPhlAn", ...)` which returns the actual database (or parts of it).
2. We can try to integrate into the current workflow, for example: queryMapping -> importMapping & Modules -> mapModules -> addModules.

Nice work! Happy to discuss the above points.

thomazbastiaanssen committed:
Removed MultiFactorDB.
Progress on method to describe which data-links are present in a database in a human & machine readable way. see `repo_info.R`

thomazbastiaanssen committed:
Implement importing interface

```
# Returns a MultiFactor, handle as usual
x <- ariadne("ChocoPhlAn")

# when ready for download:
x <- importMapping(x, dry_run = FALSE)
```

thomazbastiaanssen committed:
Cleanup.

Found a way to process modules using crossprod, sequentially.

Still work to be done before merge

thomazbastiaanssen committed:
Some updates, changed to dod.notation for dry.run

thomazbastiaanssen committed:
Add two internal functions to map modules & pathways.
`.map_ANDcomplex` returns whether all members of a complex are present in a sample.

 `.map_coverage` returns steps_present/known_steps for eacht module, i.e. module coverage.

thomazbastiaanssen committed:
Update mapModules.

thomazbastiaanssen committed:
adopting new `weave()` functionality

thomazbastiaanssen committed:
Finished push for mapModules to use MultiFactor.

mapModules now supports stratified analysis stragight from curatedMetagenomicData output.

Some work remains:
- Likely still bugs in there
- To add tests against rpm.
- Code gets complicated in some places
